src/main/java/fi/fmi/avi/model/swx/immutable/SpaceWeatherAdvisoryImpl.java

```diff
@@ -105,6 +118,62 @@ public Builder addAllPhenomena(final List<SpaceWeatherPhenomenon> elements) {
             return super.addAllPhenomena(elements);
         }
 
+        public SpaceWeatherAdvisoryImpl.Builder withCompleteIssueTimeNear(final ZonedDateTime reference) {
+            requireNonNull(reference, "reference");
+            return mapIssueTime((input) -> input.toBuilder().completePartialNear(reference).build());
+        }
+
+        private SpaceWeatherAdvisoryImpl.Builder withCompleteNextAdvisory(final ZonedDateTime reference) {
+            requireNonNull(reference, "reference");
+            return mapNextAdvisory(nextAdvisory -> {
+                final NextAdvisoryImpl.Builder builder = NextAdvisoryImpl.Builder.from(nextAdvisory);
+                builder.mapTime(time -> time.toBuilder().completePartialNear(reference).build());
+                return builder.build();
+            });
+        }
+
+        private Builder withCompleteAnalysisTimes(final ZonedDateTime reference) {
+            requireNonNull(reference, "reference");
+            if (!getAnalyses().isEmpty()) {
+                final Iterable<PartialOrCompleteTimeInstant> partialTimes = getAnalyses().stream().map(SpaceWeatherAdvisoryAnalysis::getTime)::iterator;
+                final List<PartialOrCompleteTime> times = PartialOrCompleteTimes.completeAscendingPartialTimes(partialTimes, reference,
+                        toZonedDateTimeNotBeforeOrNear());
+
+                final List<SpaceWeatherAdvisoryAnalysis> completedAnalyses = new ArrayList<>();
+                for (int i = 0; i < times.size(); i++) {
+                    final PartialOrCompleteTime time = times.get(i);
+                    completedAnalyses.add(
+                            SpaceWeatherAdvisoryAnalysisImpl.Builder.from(getAnalyses().get(i)).setTime((PartialOrCompleteTimeInstant) time).build());
```

Comment:
As before, as a minor optimization I would first check whether the completed time equals the original analysis time and build a new analysis only if times differ.